x/protorev/keeper/routes.go

```diff
@@ -0,0 +1,206 @@
+func (k Keeper) BuildRoutes(ctx sdk.Context, tokenIn, tokenOut string, poolId uint64) [][]TradeInfo {
+	routes := make([][]TradeInfo, 0)
+
+	// Check hot routes if enabled
+	if tokenPairRoutes, err := k.BuildTokenPairRoutes(ctx, tokenIn, tokenOut, poolId); err == nil {
+		routes = append(routes, tokenPairRoutes...)
+	}
+
+	// Append an osmo route if one exists
+	if osmoRoute, err := k.BuildOsmoRoute(ctx, tokenIn, tokenOut, poolId); err == nil {
+		routes = append(routes, osmoRoute)
+	}
+
+	// Append an atom route if one exists
+	if atomRoute, err := k.BuildAtomRoute(ctx, tokenIn, tokenOut, poolId); err == nil {
+		routes = append(routes, atomRoute)
+	}
+
+	return routes
+}
+
+// BuildTokenPairRoutes builds all of the possible arbitrage routes from the hot routes given the tokenIn, tokenOut and poolId that were used in the swap
+func (k Keeper) BuildTokenPairRoutes(ctx sdk.Context, tokenIn, tokenOut string, poolId uint64) ([][]TradeInfo, error) {
+	// Get all of the routes from the store that match the given tokenIn and tokenOut
+	tokenPairArbRoutes, err := k.GetTokenPairArbRoutes(ctx, tokenIn, tokenOut)
+	if err != nil {
+		return [][]TradeInfo{}, err
+	}
+
+	// Iterate through all of the routes and build hot routes
+	routes := make([][]TradeInfo, 0)
+	for _, route := range tokenPairArbRoutes.ArbRoutes {
+		newRoute := make([]TradeInfo, 0)
+
+		var newTrade TradeInfo
+		for _, trade := range route.Trades {
+			// 0 is a placeholder for swaps that should be entered into the hot route
+			if trade.Pool == 0 {
+				pool, err := k.GetAndCheckPool(ctx, poolId)
+				if err != nil {
+					return [][]TradeInfo{}, err
+				}
+
+				newTrade = TradeInfo{
+					InputDenom:  tokenOut,
+					OutputDenom: tokenIn,
+					SwapFee:     pool.GetSwapFee(ctx),
+					Pool:        pool,
+				}
+			} else {
+				pool, err := k.GetAndCheckPool(ctx, trade.Pool)
+				if err != nil {
+					return [][]TradeInfo{}, err
+				}
+
+				newTrade = TradeInfo{
+					InputDenom:  trade.TokenIn,
+					OutputDenom: trade.TokenOut,
+					SwapFee:     pool.GetSwapFee(ctx),
+					Pool:        pool,
+				}
+			}
+
+			newRoute = append(newRoute, newTrade)
+		}
+
+		// Only append if the arbitrage denom is valid
+		// In denom and out denom must be the same
+		// In denom must be uosmo or atom
+		// There must be exactly three hops in the route
+		if len(newRoute) == 3 && (newRoute[0].InputDenom == types.AtomDenomination || newRoute[0].InputDenom == types.OsmosisDenomination) && (newRoute[0].InputDenom == newRoute[2].OutputDenom) {
+			routes = append(routes, newRoute)
+		}
+	}
+
+	return routes, nil
+}
+
+// BuildOsmoRoute builds a cyclic arbitrage route that starts and ends with osmo given the tokenIn, tokenOut and poolId that were used in the swap
+func (k Keeper) BuildOsmoRoute(ctx sdk.Context, tokenIn, tokenOut string, poolId uint64) ([]TradeInfo, error) {
+	// Creating the first trade in the arb
+	entryPoolId, err := k.GetOsmoPool(ctx, tokenOut)
```

Comment:
Does this get the highest liquidity osmo pool?

Reply:
yes, it gets the highest liquidity osmo pool that has osmo <-> tokenOut as pool assets